docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+
+ 
+### Empty Profile
+
+A device profile can begin life “empty” - meaning that it has no device resources or device commands.  Once associated to any object (device, event, reading) then, since device resource or device command additions can be made, the empty profile can gain resources or commands.  You can never remove device resources or device commands once the profile is associated to another object.
+
+The general properties of a device profile, with the exception of name, are optional when creating an empty profile.
+
+All profiles – even empty profiles – must have a name.                                                                                 
+
+### Deleting Profiles
+
+A profile can only be deleted from metadata when the profile is not associated to any other EdgeX object to include device, provision watcher, event or reading.
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR
+
+- Block Device DELETE when there is any associated event/reading exists.
```

Comment:
Given that short term persistence of events/readings is less of requirements these days, does this restriction really make sense?

Reply:
As long as we do have persistence, I would maintain yes.  Otherwise you have referential integrity issues.

As an alternative, we could offer some sort of configuration option that would allow these types of delete/updates and not be concerned with references (remember we used to have a configurable check of metadata before persisting event/reading), but that seems like a lot of work for this.  

Another alternative is to offer delete of associated event/readings when you delete the associated device - but our APIs would allow a user to do this on their own.

Reply:
I wonder about enforcement, similar to Tony's question. If people implement (in their services, not the SDK) by a simple-minded, blunt approach, would it be OK? Eg, if the device service EVER published ANYTHING from the profile, it considers the profile "dirty" and won't allow the profile to be deleted, or won't allow any of it resources to be deleted?